py-polars/polars/io.py

```diff
@@ -1281,6 +1288,262 @@ def _read_excel_sheet(
     return read_csv(csv_buffer, **read_csv_options)  # type: ignore[arg-type]
 
 
+def _get_delta_lake_table(
+    table_path: str,
+    version: int | None = None,
+    storage_options: dict[str, object] | None = None,
+    delta_table_options: dict[str, object] | None = None,
+) -> deltalake.DeltaTable:
+    """
+    Initialise a Delta lake table for use in read and scan operations.
+
+    Notes
+    -----
+    Make sure to install deltalake>=0.6.0. Read the documentation
+    `here <https://delta-io.github.io/delta-rs/python/installation.html>`_.
+
+    Returns
+    -------
+    DeltaTable
+
+    """
+    if _WITH_DELTA:
+        if delta_table_options is None:
+            delta_table_options = {}
+
+        dl_tbl = deltalake.DeltaTable(
+            table_uri=table_path,
+            version=version,
+            storage_options=storage_options,
+            **delta_table_options,
+        )
+
+        return dl_tbl
+    else:
+        raise ImportError(
+            "deltalake is not installed. Please run `pip install deltalake>=0.6.0`."
+        )
+
+
+def scan_delta(
+    table_uri: str,
+    version: int | None = None,
+    raw_filesystem: pa.fs.FileSystem | None = None,
+    storage_options: dict[str, object] | None = None,
+    delta_table_options: dict[str, object] | None = None,
+    pyarrow_options: dict[str, object] | None = None,
+) -> LazyFrame:
+    """
+    Lazily read from a Delta lake table.
+
+    Parameters
+    ----------
+    table_uri
+        Path to the root directory of the Delta lake table.
+    version
+        Version of the Delta lake table.
+
+        Note: If ``version`` is not provided, latest version of delta lake
+        table is read.
+    raw_filesystem
+        A `pyarrow.fs.FileSystem` to read files from.
+
+        Note: The root of the filesystem has to be adjusted to point at the root of the
+        Delta lake table. The provided ``raw_filesystem`` is wrapped into a
+        `pyarrow.fs.SubTreeFileSystem`
+
+        More info is available `here
+        <https://delta-io.github.io/delta-rs/python/usage.html?highlight=backend#custom-storage-backends>`__.
+    storage_options
+        Extra options for the storage backends supported by `deltalake`.
+        For cloud storages, this may include configurations for authentication etc.
+
+        More info is available `here
+        <https://delta-io.github.io/delta-rs/python/usage.html?highlight=backend#loading-a-delta-table>`__.
+    delta_table_options
+        Additional keyword arguments while reading a Delta lake Table.
+    pyarrow_options
+        Keyword arguments while converting a Delta lake Table to pyarrow table.
+
+    Returns
+    -------
```

Comment:
This is a great PR. I started on something similar 2 weeks ago but my intent was to integrate at the Rust level, on top of the crates arrow2 and object-store. This turns out to be above my current async rust understanding, but I am pushing on that.

It would be great to get some guidance from the community if the python or rust layer are more interesting.

Reply:
The rust and python implementations both come from [delta-rs](https://github.com/delta-io/delta-rs/). Like Polars, the python bindings in delta depend on its rust bindings which supports async, arrow2 and parquet2 now.

So I thought that it will be trivial now to allow delta as a source directly on the python side.

@ritchie46 @stinodego any guidance/ suggestions on this ?

Reply:
I'd rather have them on the python side. It doesn't bloat our binary.

For rust users we can still support it behind a feature gate.

Reply:
cool! resolving this.